ffi/custom_passes.cpp

```diff
@@ -0,0 +1,613 @@
+                    }
+                }
+            }
+            // Remove refops on NULL
+            for (CallInst* ci: null_list) {
+                ci->eraseFromParent();
+                mutated |= true;
+                stats_per_bb += 1;
+            }
+            // Find matching pairs of incref decref
+            while (incref_list.size() > 0) {
+                CallInst* incref = incref_list.pop_back_val();
+                for (size_t i=0; i < decref_list.size(); ++i){
+                    CallInst* decref = decref_list[i];
+                    if (decref && isRelatedDecref(incref, decref)) {
+                        if (DEBUG_PRINT) {
+                            errs() << "Prune: matching pair in BB:\n";
+                            incref->dump();
+                            decref->dump();
+                            incref->getParent()->dump();
+                        }
+                        incref->eraseFromParent();
+                        decref->eraseFromParent();
+
+                        decref_list[i] = NULL;
+                        mutated |= true;
+                        stats_per_bb += 2;
+                        break;
+                    }
+                }
+            }
+        }
+        return mutated;
+    }
+
+    bool runDiamondPrune(Function &F) {
+        // Check pairs that are dominating and postdominating each other
+        bool mutated = false;
+        auto &domtree = getAnalysis<DominatorTreeWrapperPass>().getDomTree();
+        auto &postdomtree = getAnalysis<PostDominatorTreeWrapperPass>().getPostDomTree();
+
+        std::vector<CallInst*> incref_list, decref_list;
+        for (BasicBlock &bb : F) {
+            for (Instruction &ii : bb) {
+                CallInst* ci;
+                if ( (ci = GetRefOpCall(&ii)) ) {
+                    if ( IsIncRef(ci) ) {
+                        incref_list.push_back(ci);
+                    }
+                    else if ( IsDecRef(ci) ) {
+                        decref_list.push_back(ci);
+                    }
+                }
+            }
+        }
+
+        bool diamond = false;
+        for (CallInst*& incref: incref_list) {
+            if (incref == NULL) continue;
+
+            for (CallInst*& decref: decref_list) {
+                if (decref == NULL) continue;
+
+                // Not the same BB
+                if (incref->getParent() == decref->getParent() ) continue;
+
+                // Is related refop pair
+                if (!isRelatedDecref(incref, decref)) continue;
+
+                // incref DOM decref && decref POSTDOM incref
+                if ( domtree.dominates(incref, decref)
+                        && postdomtree.dominates(decref, incref) ){
+                    if (incref->getParent() != decref->getParent() ) {
+                        SmallVector<BasicBlock*, 20> stack;
+                        if (hasDecrefBetweenGraph(incref->getParent(), decref->getParent(), stack)) {
```

Comment:
This is discovered from a failed test.  My current explanation is at https://hackmd.io/LJLRmcKdTgeP5RjA0aOxOw#Problem-1

Reply:
Thanks for the explanation - how come normalizing reference counts didn't move the decref to after the incref? Is the example simplified, in that C is a placeholder for multiple basic blocks so that the normalization couldn't move the decref after the incref?

Reply:
Correct, a simplified example. Let me go update the example